### ISSUE TITLE
Question: postgres-atomicity

### ISSUE THREAD

**Issue by nico1510** [brandur/sorg#151]
Hey first of all thank you for this great blog post here https://brandur.org/postgres-atomicity. I have a quick question about the visibility section. From what you wrote the `HeapTupleSatisfiesVisibility` method seems like it's just checking if the tuple's `xmin` is contained in the commit log. Am I right in assuming that it only does that if the following conditions apply ?
1) the tuple's `xmin` is greater than or equal to the snapshot's `xmin`
2) the tuple's `xmin` is *not* conainted in the snapshot's `xip` array
3) the tuple's `xmin` is smaller than the snapshot's `xmax`

In your example:
![Image of Yaktocat](https://brandur.org/assets/postgres-atomicity/snapshot-creation.svg)
I would have thought that you would only need consult the commit log if a tuple has the transaction ID 126 as its `xmin`. In all other cases visibility could also be computed without checking if the transaction was commited. Did I misunderstand something here ?

**Comment by brandur** [comment]
Hi @nico1510,

Yes, what you wrote is correct. I truncated some code in the blog post because I was trying to simplify a very complex story as much as possible, but check out this version that's been slightly expanded compared to what I have there:

``` c
bool
HeapTupleSatisfiesMVCC(HeapTuple htup, Snapshot snapshot,
					   Buffer buffer)
{
...

		else if (XidInMVCCSnapshot(HeapTupleHeaderGetRawXmin(tuple), snapshot))
			return false;
		else if (TransactionIdDidCommit(HeapTupleHeaderGetRawXmin(tuple)))
			SetHintBits(tuple, buffer, HEAP_XMIN_COMMITTED,
						HeapTupleHeaderGetRawXmin(tuple));

...
}
```

Notice that there's a conditional clause _before_ we check `TransactionIdDidCommit`! I'd encourage you to take a look at the implementation of `XidInMVCCSnapshot`, but it does pretty much exactly what you'd expect — it checks whether we even need to consider this XID at all given our current snapshot, and if not, we just `return false` and skip checking the clog.

Hopefully that makes sense. Keep in mind too that code will make a better canonical source of truth here than I will :)

**Comment by brandur** [comment]
And just for ease-of-reading, here's a simplified version of `XidInMVCCSnapshot` where you can see checks against each of `xmin`, `xmax`, and `xip`:

``` c
static bool
XidInMVCCSnapshot(TransactionId xid, Snapshot snapshot)
{
	uint32		i;

	/*
	 * Make a quick range check to eliminate most XIDs without looking at the
	 * xip arrays.  Note that this is OK even if we convert a subxact XID to
	 * its parent below, because a subxact with XID < xmin has surely also got
	 * a parent with XID < xmin, while one with XID >= xmax must belong to a
	 * parent that was not yet committed at the time of this snapshot.
	 */

	/* Any xid < xmin is not in-progress */
	if (TransactionIdPrecedes(xid, snapshot->xmin))
		return false;
	/* Any xid >= xmax is in-progress */
	if (TransactionIdFollowsOrEquals(xid, snapshot->xmax))
		return true;

	...

	for (i = 0; i < snapshot->xcnt; i++)
	{
		if (TransactionIdEquals(xid, snapshot->xip[i]))
			return true;
	}

	...
}
```

**Comment by nico1510** [comment]
Ahh yes thank you very much @brandur !! The Postgres source code is a little overwhelming to look at but your extracts make it pretty obvious what's happening. I think it's a fantastic idea to truncate the original source code and show the important bits.
As an idea for improvement (feel free to ignore) I would add a sentence or two about how `xmin`, `xmax` and `xip` are used for checking visibility. Because the #snapshots sections talks about how these values are being constructed and in the end it seems like they are not used anywhere. But maybe that's just me.
Thanks again ;)

**Comment by brandur** [comment]
Thanks for the feedback. I get the feeling that most people did not read far enough to have noticed that omission ;)

I just merged #153 which contains an addition roughly the same as what's described here.

**Comment by nico1510** [comment]
perfect :-)
